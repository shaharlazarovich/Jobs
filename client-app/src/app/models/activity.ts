export interface IAttendee {
    userName: string;
    displayName: string;
    image: string;
    isHost: boolean;
    following?: boolean,
}

//we are adding this interface to mimic what we do on the server for paging -
//which is to have an entity around our activity array, in order to add a count
//of activities - which then, we could device in the number of pages
export interface IActivitiesEnvelope {
    activities: IActivity[];
    activityCount: number
}

export interface IActivity {
    id: string;
    title: string;
    description: string;
    category: string;
    date: Date;
    city: string;
    venue: string;
    isGoing: boolean;
    isHost: boolean;
    attendees: IAttendee[];
    comments: IComment[];
}

export interface IComment {
    id: string;
    createdAt: Date;
    body: string;
    username: string;
    displayName: string;
    image: string;
}

export interface IActivityFormValues extends Partial<IActivity> {
    time?: Date    
}

export class ActivityFormValues implements IActivityFormValues {
    id?: string = undefined;
    title: string = '';
    category: string = '';
    description: string = '';
    date?:Date = undefined;
    time?: Date = undefined;
    city: string = '';
    venue: string = '';

    constructor(init?: IActivityFormValues) {
        if (init && init.date) {
            init.time = init.date;
        }
        Object.assign(this, init); //this will initialize our object of class ActivityFormValues
                                  // with the above values
    }
}