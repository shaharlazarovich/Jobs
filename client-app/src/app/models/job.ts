//we are adding this interface to mimic what we do on the server for paging -
//which is to have an entity around our jobs array, in order to add a count
//of jobs - which then, we could device in the number of pages
export interface IJobsEnvelope {
    jobs: IJob[];
    jobsCount: number
}

export interface IJob {
    id: number;
    jobName: string;
    company: string;
    replication: string;
    servers: string;
    lastRun: Date;
    date: Date;
    time: Date;
    rta: string;
    results: string;
    key: string;
    rtoNeeded: string;
    jobIP: string;
}


export interface IJobFormValues extends Partial<IJob> {
     
}

export class JobFormValues implements IJobFormValues {
    id?: number = 0;
    jobName: string = '';
    company: string = '';
    replication: string = '';
    servers: string = '';
    lastRun?:Date = undefined;
    date?:Date = undefined;
    time?:Date = undefined;
    rta?: string = '';
    results: string = '';
    key: string = '';
    rtoNeeded: string = '';
    jobIP: string = '';

    constructor(init?: IJobFormValues) {
        Object.assign(this, init); //this will initialize our object of class JobFormValues
                                  // with the above values
    }
}