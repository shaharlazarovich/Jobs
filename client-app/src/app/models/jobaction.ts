//we are adding this interface to mimic what we do on the server for paging -
//which is to have an entity around our jobs array, in order to add a count
//of jobs - which then, we could device in the number of pages
export interface IJobActionsEnvelope {
    jobActions: IJobAction[];
    jobActionsCount: number
}

export interface IJobAction {
    id: string;
    jobId: string;
    jobName: string;
    userId: string;
    actionDate: Date;
    remoteIP: string;
    remoteResponse: string;
    requestProperties: string;
    source: string;
    action: string;
}

export class JobAction implements IJobAction {
    id: string = '';
    jobId: string = '';
    jobName: string = '';
    userId: string = '';
    actionDate: Date = new Date();
    remoteIP: string = '';
    remoteResponse: string = '';
    requestProperties: string = '';
    source: string = '';
    action: string = '';

    constructor(init?: IJobAction) {
        Object.assign(this, init); //this will initialize our object of class JobActions
                                  // with the above values
    }
}