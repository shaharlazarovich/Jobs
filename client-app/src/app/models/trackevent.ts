//we are adding this interface to mimic what we do on the server for paging -
//which is to have an entity around our jobs array, in order to add a count
//of jobs - which then, we could device in the number of pages
export interface ITrackEventsEnvelope {
    trackEvents: ITrackEvent[];
    trackEventsCount: number
}

export interface ITrackEvent {
    id: string;
    jobId?: string;
    jobName?: string;
    userId?: string;
    actionDate: Date;
    remoteIP?: string;
    remoteResponse?: string;
    requestProperties?: string;
    source: string;
    event: string;
}

export class TrackEvent implements ITrackEvent {
    id: string = '';
    jobId?: string = '';
    jobName?: string = '';
    userId?: string = '';
    actionDate: Date = new Date();
    remoteIP?: string = '';
    remoteResponse?: string = '';
    requestProperties?: string = '';
    source: string = '';
    event: string = '';

    constructor(init?: ITrackEvent) {
        Object.assign(this, init); //this will initialize our object of class JobActions
                                  // with the above values
    }
}