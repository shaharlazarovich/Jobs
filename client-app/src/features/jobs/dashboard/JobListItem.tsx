import React, { useContext } from 'react'
import { Item, Button, Segment, Icon } from 'semantic-ui-react'
import { Link } from 'react-router-dom'
import { IJob } from '../../../app/models/job'
import { TrackEvent } from '../../../app/models/trackevent'
import {format} from 'date-fns'
import { RootStoreContext } from '../../../app/stores/rootStore'
import moment from 'moment'
import {v4 as uuid} from 'uuid'

const JobListItem: React.FC<{ job: IJob }> = ({ job }) => {
  const rootStore = useContext(RootStoreContext);
  const { user} = rootStore.userStore;
  const { runRemoteJob } = rootStore.jobStore;

  const track = () => {
    let trackEvent = new TrackEvent();
    trackEvent = {
      id: uuid(),
      jobId: job.id,
      jobName: job.jobName,
      userId: user!.username,
      actionDate: moment().toDate(),
      remoteIP: job.jobIP,
      remoteResponse: "",
      requestProperties: "",
      source: "Jobs_List",
      event: "runjob",
    }
    runRemoteJob(trackEvent);
  }

  return (
    <Segment.Group data-test="component-job-list-item">
      <Segment>
          <Item.Group>
          <Item>
          <Item.Image size='tiny' circular src={'/assets/user.png'}
          style={{marginBottom: 3}} />
          <Item.Content>
            <Item.Header as={Link} to={`/jobs/${job.id}`}>{job.jobName}</Item.Header>
            <Item.Description>
              Created by <Link to={`/profile/${user!.username}`}> {user!.displayName }</Link>
            </Item.Description>
          </Item.Content>
        </Item>
        </Item.Group>
      </Segment>
      <Segment>
          <Icon name='clock' />{format(job.lastRun, 'h:mm a')}
          <Icon name='marker' />{job.rtoNeeded} , {job.rta}
      </Segment>
      <Segment clearing>
            <span>{job.results}</span>
            <Button
                as={Link}
                to={`/jobs/${job.id}`}
                floated="right"
                content="View"
                color="blue"
            />
            <Button
                onClick={ track }
                floated="right"
                content="Run"
                color="blue"
            />
      </Segment>
    </Segment.Group>
  );
};

export default JobListItem
