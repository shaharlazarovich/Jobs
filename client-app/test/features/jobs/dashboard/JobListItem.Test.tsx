import { shallow } from "enzyme";
import React from "react";
import { findByTestAttr } from '../../../common/testUtils';
import { IJob, JobFormValues } from '../../../../src/app/models/job'
import JobListItem from "../../../../src/features/jobs/dashboard/JobListItem";

const setup = (job:any) => {
    const wrapper = shallow<IJob>(<JobListItem job={job} />)
    return wrapper;
}

describe('Job List Item UniTest', () => {
    let wrapper:any;
    let testJob = new JobFormValues();
    beforeEach(() => {
        testJob = { 
            id:"4", 
            replication:"VMware SRM",
            jobName:"test1",
            company: 'Shomera',  
            lastRun: new Date('2020-01-06 23:00:05'),  
            servers: '10',
            results: 'OK',
            key: 'AAAA-BBBB-CCCC-DDDD',
            rtoNeeded: "10"
        }
        wrapper = setup({ testJob });
    });
    test('renders without error', () => {
        const component = findByTestAttr(wrapper, 'component-job-list-item');
        expect(component.length).toBe(1);
    });
})