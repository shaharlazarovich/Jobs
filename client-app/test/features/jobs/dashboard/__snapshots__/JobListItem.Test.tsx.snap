// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Job List Item UniTest should match the snapshot 1`] = `
<SegmentGroup
  data-test="component-job-list-item"
>
  <div
    className="ui segments"
    data-test="component-job-list-item"
  >
    <Segment>
      <div
        className="ui segment"
      >
        <ItemGroup>
          <div
            className="ui items"
          >
            <Item>
              <div
                className="item"
              >
                <ItemImage
                  circular={true}
                  size="tiny"
                  src="/assets/user.png"
                  style={
                    Object {
                      "marginBottom": 3,
                    }
                  }
                >
                  <Image
                    as="img"
                    circular={true}
                    size="tiny"
                    src="/assets/user.png"
                    style={
                      Object {
                        "marginBottom": 3,
                      }
                    }
                    ui={true}
                    wrapped={true}
                  >
                    <div
                      className="ui tiny circular image"
                      style={
                        Object {
                          "marginBottom": 3,
                        }
                      }
                    >
                      <img
                        src="/assets/user.png"
                      />
                    </div>
                  </Image>
                </ItemImage>
                <ItemContent>
                  <div
                    className="content"
                  >
                    <ItemHeader
                      as={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "displayName": "Link",
                          "propTypes": Object {
                            "innerRef": [Function],
                            "onClick": [Function],
                            "replace": [Function],
                            "target": [Function],
                            "to": [Function],
                          },
                          "render": [Function],
                        }
                      }
                      to="/jobs/4"
                    >
                      <Link
                        className="header"
                        to="/jobs/4"
                      >
                        <LinkAnchor
                          className="header"
                          href="/jobs/4"
                          navigate={[Function]}
                        >
                          <a
                            className="header"
                            href="/jobs/4"
                            onClick={[Function]}
                          >
                            test1
                          </a>
                        </LinkAnchor>
                      </Link>
                    </ItemHeader>
                    <ItemDescription>
                      <div
                        className="description"
                      >
                        Created by 
                        <Link
                          to="/profile/Fake Username"
                        >
                          <LinkAnchor
                            href="/profile/Fake Username"
                            navigate={[Function]}
                          >
                            <a
                              href="/profile/Fake Username"
                              onClick={[Function]}
                            >
                               
                              Fake Username
                            </a>
                          </LinkAnchor>
                        </Link>
                      </div>
                    </ItemDescription>
                  </div>
                </ItemContent>
              </div>
            </Item>
          </div>
        </ItemGroup>
      </div>
    </Segment>
    <Segment>
      <div
        className="ui segment"
      >
        <Icon
          as="i"
          name="clock"
        >
          <i
            aria-hidden="true"
            className="clock icon"
            onClick={[Function]}
          />
        </Icon>
        11:00 PM
        <Icon
          as="i"
          name="marker"
        >
          <i
            aria-hidden="true"
            className="marker icon"
            onClick={[Function]}
          />
        </Icon>
        10
         , 
        10
      </div>
    </Segment>
    <Segment
      clearing={true}
    >
      <div
        className="ui clearing segment"
      >
        <span>
          OK
        </span>
        <Button
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "Link",
              "propTypes": Object {
                "innerRef": [Function],
                "onClick": [Function],
                "replace": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          color="blue"
          content="View"
          floated="right"
          to="/jobs/4"
        >
          <Ref
            innerRef={
              Object {
                "current": <a
                  class="ui blue right floated button"
                  href="/jobs/4"
                  role="button"
                >
                  View
                </a>,
              }
            }
          >
            <RefForward
              innerRef={
                Object {
                  "current": <a
                    class="ui blue right floated button"
                    href="/jobs/4"
                    role="button"
                  >
                    View
                  </a>,
                }
              }
            >
              <Link
                className="ui blue right floated button"
                onClick={[Function]}
                role="button"
                to="/jobs/4"
              >
                <LinkAnchor
                  className="ui blue right floated button"
                  href="/jobs/4"
                  navigate={[Function]}
                  onClick={[Function]}
                  role="button"
                >
                  <a
                    className="ui blue right floated button"
                    href="/jobs/4"
                    onClick={[Function]}
                    role="button"
                  >
                    View
                  </a>
                </LinkAnchor>
              </Link>
            </RefForward>
          </Ref>
        </Button>
        <Button
          as="button"
          color="blue"
          content="Run"
          floated="right"
          onClick={[Function]}
        >
          <Ref
            innerRef={
              Object {
                "current": <button
                  class="ui blue right floated button"
                >
                  Run
                </button>,
              }
            }
          >
            <RefFindNode
              innerRef={
                Object {
                  "current": <button
                    class="ui blue right floated button"
                  >
                    Run
                  </button>,
                }
              }
            >
              <button
                className="ui blue right floated button"
                onClick={[Function]}
              >
                Run
              </button>
            </RefFindNode>
          </Ref>
        </Button>
      </div>
    </Segment>
  </div>
</SegmentGroup>
`;
